File: Curso Programação Orientada a Objetos/PHP com POO/Aula12/Cachorro.php
<?php

require_once 'Mamifero.php';

class Cachorro extends Mamifero {
    //Métodos
    public function enterrarOsso() {
        echo "<p>Cachorro está enterrando seu osso!</p>";
    }
    
    public function abanarRabo() {
        echo "<p>Cachorro está abanando seu rabo!</p>";
    }
    
    //Métodos Sobrepostos
    public function emitirSom() {
        echo "<p>Au! Au! Au!</p>";
    }
    
    //Testando Métodos de Polimorfismo de Sobrecarga no PHP -> Aula 13
    
    //public function reagir($frase) {
    
    //public function reagir($hora, $min) {
    
    //public function reagir($dono) {
    
    //public function reagir($idade, $peso) {
    
    //PHP não suporta Polimorfismo de Sobrecarga
    //Não é possível declarar métodos com o mesmo nome, mesmo que as assinaturas sejam diferentes. 
    
    //A alternativa para resolver isso é: 
    //Declarar métodos de nomes parecidos, mas não iguais. 
    
    public function reagirFrase($frase) {
        if ($frase == "Toma Cuidado" || $frase == "Olá") {
            echo "<p>Abanar e Latir</p>";
        } else {
            echo "<p>Rosnar</p>";
        }
    }
    
    public function reagirHorario($hora, $min) {
        if ($hora < 12) {
            echo "<p>Abanar</p>";
        } elseif ($hora >= 18) {
            echo "<p>Ignorar</p>";
        } else {
            echo "<p>Abanar e Latir</p>";
        }
    }
    
    public function reagirDono($dono) {
        if ($dono == true) {
            echo "<p>Abanar</p>";
        } else {
            echo "<p>Rosnar e Latir</p>";
        }
    }
    
    public function reagirIdadePeso($idade, $peso) {
        if ($idade < 5) {
            if ($peso < 10) {
                echo "<p>Abanar</p>";
            } else {
                echo "<p>Latir</p>";
            }
        } else {
            if ($peso < 10) {
                echo "<p>Rosnar</p>";
            } else {
                echo "<p>Ignorar</p>";
            }
        }
    }
}

File: Curso Programação Orientada a Objetos/Java com POO/Aula12e13/src/aula12e13/Cachorro.java
package aula12e13;

public class Cachorro extends Mamifero {
    //Métodos
    public void enterrarOsso() {
        System.out.println("Cachorro está enterrando o osso!");
    }
    
    public void abanarRabo() {
        System.out.println("Cachorro está abanando o rabo!");
    }
    
    //Métodos Sobrepostos
    @Override
    public void emitirSom() {
        System.out.println("Au! Au! Au!");
    }
    
    //Aula 13 -> Polimorfismo de Sobrecarga
    //Novos métodos criados para o exemplo:
    //Método reagir implementados 4 vezes com assinaturas diferentes e mesma classe
    //1
    public void reagir(String frase) { 
        //if (frase == "Toma cuidado" || frase == "Olá") {
        if (frase.equals("Toma Cuidado") || frase.equals("Olá")) {
            System.out.println("Abanar e Latir");
        } else {
            System.out.println("Rosnar");
        }
    }
    
    //2
    public void reagir(int hora, int min) { 
        if (hora < 12) {
            System.out.println("Abanar");
        } else if (hora >= 18) {
            System.out.println("Ignorar");
        } else {
            System.out.println("Abanar e Latir");
        }
    }
    
    //3
    public void reagir(boolean dono) { 
        if (dono == true) {
            System.out.println("Abanar");
        } else {
            System.out.println("Rosnar e Latir");
        }
    }
    
    //4
    public void reagir(int idade, float peso) { 
        if (idade < 5) {
            if (peso < 10) {
                System.out.println("Abanar");
            } else {
                System.out.println("Latir");
            }
        } else {
            if (peso < 10) {
                System.out.println("Rosnar");
            } else {
                System.out.println("Ignorar");
            }
        }
    }
}
